Index changes for log segmentation

vgkholla opened the pull request:
This change introduces changes in the index to handle multiple log segments.
The index is now aware of the log being segmented and deals with all operations accordingly.
Hard delete can also deal with multiple log segments now (by virtue of using the index)

`./gradlew build && ./gradlew test` passed

Most of the changes are in `PersistentIndex` and its tests.

The current tests succeed which means that all of this code works for the single segment use case (the current status quo). After talking with the reviewers, we have agreed that we will continue to merge PRs as long as all the current tests succeed and the reviewers are satisfied that it works with the multiple log segment use case. Once all the relevant PRs are merged, I will make a commit that contains unit and integration tests that exercise the multiple log segment use cases to increase confidence in the code and fix bugs that arise, if any.

**Primary reviewers: @pnarayanan, @nsivabalan**

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -453,19 +449,19 @@ public IndexValue findKey(StoreKey key, FileSpan fileSpan) throws StoreException
    */
-  public void markAsDeleted(StoreKey id, FileSpan fileSpan) throws StoreException {
-    validateFileSpan(fileSpan);
+  void markAsDeleted(StoreKey id, FileSpan fileSpan) throws StoreException {
```
can you enlighten me as to how we set the original msg offset for a delete record

vgkholla replied:
It is in the `setNewOffset` method of `IndexValue`.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -167,27 +166,27 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
-        // TODO (Index changes): Working under the assumption that there is only one log segment.
-        long currentLogEndOffset = log.getEndOffset().getOffset();
+        Offset currentLogEndOffset = log.getEndOffset();
         messageSetToWrite.writeTo(log);
```
This makes me a little uncomfortable. I feel that the `IndexValue`s should be created in a more tightly coupled way. There is a bit of an abstraction leakage here in that the `Log` determines where the individual messages go and then the `Store` sort of "figures" out the offsets of these messages in the log in creating the `IndexValue`s - and does not even consult the `Log` for that.

I understand that the old code did it this way as well - but at least in the old code it was a simple addition. Now that the offset additions are more sophisticated, for one I think we should let just one class deal with that. Secondly, since we made the change to write each message in the `writeSet` individually to the `Log` - perhaps the `IndexValue`s can be created after each individual message? Perhaps the `writeSet` could expose an iterator-like interface:
```java
while (writeSet.hasNext()) {
  writeSet.writeNext(); // or some such thing.
  ... generate the `Offset` of the last message with help from the Log.
}
```

vgkholla replied:
I agree with you and it makes me uncomfortable too. But unless the `writeSet` has a way of communicating this, I don't think we can make it any better.

I agree with having it one class and will move `getFileSpanForMessage()` to the `Log`. 

Regarding the interface change you suggest, it can be explored but since it involves touching interfaces outside the store package - if we are going to do it, we should do it another PR.  Also, I'm not sure what the interface change would buy us. Generating one by one as they are written v/s generating after everything is written should be no different in terms of complexity/bugs etc right?

vgkholla replied:
I have added comprehensive tests for the `getFileSpanForMessage()` method and i'm sure that works without bugs. Let me know if you think of any test cases and I can confirm whether it is covered or add a new test. Will eventually add test for `BlobStore` also in the tests PR that we have discussed.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -167,27 +166,27 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
         for (MessageInfo info : messageInfo) {
+          FileSpan fileSpan = getFileSpanForMessage(currentLogEndOffset, info.getSize());
           IndexValue value =
               new IndexValue(info.getSize(), currentLogEndOffset, (byte) 0, info.getExpirationTimeInMs());
```
I think this is wrong. `currentLogEndOffset` is not necessarily the start of the message. This should be:
```java
new IndexValue(info.getSize(), fileSpan.getStartOffset(), (byte) 0, info.getExpirationTimeInMs());
```

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -167,27 +166,27 @@ public void put(MessageWriteSet messageSetToWrite) throws StoreException {
           IndexEntry entry = new IndexEntry(info.getStoreKey(), value);
           indexEntries.add(entry);
-          currentLogEndOffset += info.getSize();
+          currentLogEndOffset = fileSpan.getEndOffset();
```
`currentLogEndOffset` could be renamed to `endOffsetOfLastMessage` or something - in the loop that makes more sense (as the offset is not really the "current" end offset in the log).

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -13,96 +13,127 @@
     ByteBuffer bufWrap = ByteBuffer.wrap(buf);
     bufWrap.putShort(version);
-    bufWrap.putLong(offset);
+    bufWrap.putLong(offset.getOffset());
```
Sorry, I did not understand the comment. What do you mean by fixing?

nsivabalan replied:
for version 0, we have a long at this position. but for version 1, we will have an object of "Offset" class right. So this might need fixing or am i missing something

vgkholla replied:
This will be done in a separate PR.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -13,96 +13,127 @@
 
-  BlobReadOptions(long offset, long size, long ttl, StoreKey storeKey) {
+  BlobReadOptions(Log log, Offset offset, long size, long ttl, StoreKey storeKey) {
+    segment = log.getSegment(offset.getName());
```
That's fine - that would be a programming error i.e the index had an entry that referred to no log segment. In that case, `NullPointerException` is as good as any. 

Did you have something else in mind?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -13,96 +13,127 @@
       case 0:
-        long offset = stream.readLong();
+        // backwards compatibility
+        Offset offset = new Offset(log.getStartOffset().getName(), stream.readLong());
```
done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -112,23 +143,11 @@ public static BlobReadOptions fromBytes(DataInputStream stream, StoreKeyFactory
+  StoreMessageReadSet(List<BlobReadOptions> readOptions) {
     Collections.sort(readOptions);
-    for (BlobReadOptions readOption : readOptions) {
-      if (!readOption.validateFileEndOffset(fileEndPosition)) {
```
Validation happens in the constructor of `BlobReadOptions`

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java:
```diff
@@ -205,7 +207,7 @@ void performRecovery()
   void pruneHardDeleteRecoveryRange() {
     StoreFindToken logFlushedTillToken = (StoreFindToken) startTokenSafeToPersist;
-    if (logFlushedTillToken != null && !logFlushedTillToken.isUninitialized()) {
+    if (logFlushedTillToken != null && !logFlushedTillToken.getType().equals(StoreFindToken.Type.Uninitialized)) {
```
Because there is an equivalent check - the method wasn't adding value like it was before.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -335,8 +343,13 @@ private int findIndex(StoreKey keyToFind, ByteBuffer mmap)
+    if (!entry.getValue().getOffset().getName().equals(getLogSegmentName())) {
+      throw new IllegalArgumentException("Log segment referred to by the index value [" +
+          entry.getValue().getOffset().getName() + "] is not the same as the one referred to by this index segment ["
+          + indexFile.getName() + "]");
```
It's the same since the index file name has the name of the log segment within it but i have removed this check for now since I noticed there are no checks to see if key size and value size match.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -461,14 +428,17 @@ public int getNumberOfItems() {
-    if (prevSegmentEndOffset != safeEndPoint) {
-      if (safeEndPoint > getEndOffset()) {
+    if (safeEndPoint.compareTo(startOffset) < 0) {
+      return;
```
Not sure why we need to throw? Seems like a benign situation.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -484,11 +454,11 @@ public void writeIndexToFile(long safeEndPoint)
         // write key, value size and file end pointer for this index
         writer.writeInt(this.keySize);
```
The name of the index file has that information (like start offset)

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -484,11 +454,11 @@ public void writeIndexToFile(long safeEndPoint)
         // write key, value size and file end pointer for this index
         writer.writeInt(this.keySize);
         writer.writeInt(this.valueSize);
-        writer.writeLong(safeEndPoint);
```
Not sure what you mean. That version _is_ the version for the `IndexSegment`. There cannot be a concept of "version" for the `PersistentIndex` right?

nsivabalan replied:
got you

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -583,25 +553,26 @@ private void readFromFile(File fileToRead, Journal journal)
+                  && blobValue.getOriginalMessageOffset() >= startOffset.getOffset()) {
                 // we add an entry for the original message offset if it is within the same index segment
-                journal.addEntry(blobValue.getOriginalMessageOffset(), key);
+                journal.addEntry(new Offset(startOffset.getName(), blobValue.getOriginalMessageOffset()), key);
```
I don't see a reason to do it either way...

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -699,5 +670,42 @@ public boolean getEntriesSince(StoreKey key, FindEntriesCondition findEntriesCon
+      startOffsetValue = filename.substring(0, firstSepIdx);
+    } else {
+      // pattern is logSegmentName_offset_index.
+      int lastButOneSepIdx = filename.substring(0, lastSepIdx).lastIndexOf(BlobStore.SEPARATOR);
```
`logSegmentName` can have (and has) "_" separators too.

nsivabalan replied:
couldn't get you. as per comments, pattern is either "offset_index" or "logSegmentName_offset_index". Is there any other possible pattern ?

vgkholla replied:
logSegmentName itself contains "_". An example log segment name is "0_0". So the index segment that starts at offset 0 will be named "0_0_0_index".

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -126,75 +135,54 @@ public PersistentIndex(String datadir, ScheduledExecutorService scheduler, Log l
-          map = true;
-        }
+        boolean map = i < indexFiles.length - 1;
         IndexSegment info = new IndexSegment(indexFiles[i], map, factory, config, metrics, journal);
```
Trying to keep unrelated changes out of this PR - already quite large. The index segment is inconsistently named across the `PersistentIndex`. If we are concerned about consistency, we should make those changes as a separate commit.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -566,12 +612,12 @@ public FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) th
+          offsetToStart = log.getStartOffset();
           inclusive = true;
         } else if (tokenWasReset) {
           offsetToStart = logEndOffsetOnStartup;
```
Take a look at the code a few lines above where the value of `tokenWasReset` is set. It is set to `true` if the offset in the token is > `logEndOffsetOnStartup`. So `offsetToStart = logEndOffsetOnStartup` is correct.

vgkholla replied:
I see what you mean. Will simplify

vgkholla replied:
done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -338,14 +336,24 @@ private void checkStarted() throws StoreException {
-    LogSegment firstSegment = log.getFirstSegment();
-    firstSegment.setEndOffset(index.getCurrentEndOffset());
-    log.setActiveSegment(firstSegment.getName());
+  private FileSpan getFileSpanForMessage(Offset writeStartOffset, long size) {
```
moved.

vgkholla commented:
@nsivabalan @pnarayanan Have addressed comments. Formatted and built (build and test) successfully.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -456,13 +420,16 @@ public int getNumberOfItems() {
+    if (safeEndPoint.compareTo(startOffset) < 0) {
+      return;
+    }
+    if (!safeEndPoint.equals(prevSegmentEndOffset)) {
```
what really is `prevSegmentEndOffset`? I am pretty sure it is badly and confusingly named. For a while I thought this referred to an offset in the previous segment. 
Seems like it is used to figure out whether there have been any entries since the last time this file was persisted. This should probably be called `prevSafeEndPoint` or `persistedTillOffset` or something. This change I think should go in this PR itself.

vgkholla replied:
Will change.

vgkholla replied:
Done

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -479,11 +446,11 @@ public void writeIndexToFile(long safeEndPoint) throws IOException, StoreExcepti
         // write the entries
         for (Map.Entry<StoreKey, IndexValue> entry : index.entrySet()) {
-          if (entry.getValue().getOffset() + entry.getValue().getSize() <= safeEndPoint) {
+          if (entry.getValue().getOffset().getOffset() + entry.getValue().getSize() <= safeEndPoint.getOffset()) {
```
it is guaranteed that index segments do not span log segments, right?

vgkholla replied:
Yes.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -567,35 +534,36 @@ private void readFromFile(File fileToRead, Journal journal) throws StoreExceptio
+                maxEndOffset = offsetInLogSegment + blobValue.getSize();
               }
             } else {
               logger.info(
```
Given how `writeToFile()` is done, I think the `else` case here will never happen unless there is a programming error - and if it happens it should be logged as an error and not info. Do you think there are valid cases when this can happen?

vgkholla replied:
There might have been a historical reason to do it which might not be true anymore - I'm not comfortable changing this code because it is unrelated to my changes.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -567,35 +534,36 @@ private void readFromFile(File fileToRead, Journal journal) throws StoreExceptio
               // add to the journal
-              if (blobValue.getOffset() != blobValue.getOriginalMessageOffset()
-                  && blobValue.getOriginalMessageOffset() >= startOffset.get()) {
+              if (offsetInLogSegment != blobValue.getOriginalMessageOffset()
```
Original message offset can now be of 3 types: same as the record offset, offset of the original record, or -1 (which represents invalid). Is that correct?
If so, I feel that you should have a specific check for `!= -1` and not depend on `> = startOffset.getOffset()` to take care of that case (as -1 is simply represents INVALID here).

vgkholla replied:
Will do.

vgkholla replied:
Done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -76,33 +66,43 @@
+  private final String indexBloomFilenamePrefix;
+  private final Offset startOffset;
+  private final AtomicReference<Offset> endOffset;
+  private final File indexFile;
```
Sure. Yes - some of the names are confusing.

vgkholla replied:
Created an issue for this https://github.com/linkedin/ambry/issues/499

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -612,20 +656,15 @@ public FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) th
               (time.milliseconds() - startTimeInMs));
 
           StoreFindToken storeFindToken = new StoreFindToken(offsetEnd, sessionId);
-          if (messageEntries.size() == 0) {
```
why removed this

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -125,73 +141,52 @@ public PersistentIndex(String datadir, ScheduledExecutorService scheduler, Log l
+      this.maxInMemoryIndexSizeInBytes = config.storeIndexMaxMemorySizeBytes;
+      this.maxInMemoryNumElements = config.storeIndexMaxNumberOfInmemElements;
 
+      logAbsoluteZeroOffset = new Offset(log.getFirstSegment().getName(), 0);
```
I know you might have thought about this, but may I know the reason to choose 0 as startOffset and not the startOffset of the first segment. Lets say not writes have gone into the store and if someone tries to find the totalBytesRead, wouldn't we return startOffset - 0, rather than 0. Because in find the total bytes read as part of findEntriesSince(), we calculate the diff from 0 and not startOffset of the first segment. Or a I missing something.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -341,11 +350,11 @@ public void addEntry(IndexEntry entry, long fileEndOffset) throws StoreException
       if (index.put(entry.getKey(), entry.getValue()) == null) {
         numberOfItems.incrementAndGet();
-        sizeWritten.addAndGet(entry.getKey().sizeInBytes() + IndexValue.Index_Value_Size_In_Bytes);
+        sizeWritten.addAndGet(entry.getKey().sizeInBytes() + entry.getValue().getBytes().capacity());
```
The `IndexSegment` in theory should be able to handle any value size (the value size is written into the header. Using absolute values is against that feature.

pnarayanan replied:
A given `IndexSegment` in theory is *not* designed to handle any value size. I think you meant across segments. Within a segment it has to be the same for the binary search to work correctly. We do ensure that a rollover happens whenever the value size changes, but that code is not well written right now (may not even be correct). If we set about making this change I think you should get rid of this static variable (`Index_Value_Size_In_Bytes`) altogether. What does it really represent anyway? It seems like this should be a segment specific variable.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -456,13 +420,16 @@ public int getNumberOfItems() {
    *  crc             - the crc of the index segment content
    *
-   * @param safeEndPoint
+   * @param safeEndPoint the end point until which it is safe to flush
```
Will change.

vgkholla replied:
Done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -672,24 +643,63 @@ public boolean getEntriesSince(StoreKey key, FindEntriesCondition findEntriesCon
+  /**
+   * @return the prefix for the index and bloom file names.
+   */
+  private String generateIndexBloomFilenamePrefix() {
```
Will change.

vgkholla replied:
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -672,24 +643,63 @@ public boolean getEntriesSince(StoreKey key, FindEntriesCondition findEntriesCon
+   * @return the start {@link Offset} in the {@link Log} that the index with file name {@code filename} represents.
+   */
+  static Offset getIndexSegmentStartOffset(String filename) {
+    // file name pattern for index is logSegmentName_offset_index.
```
Will change.

vgkholla replied:
Done.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -49,40 +49,56 @@
-  // TODO (Index Changes): This will stay until the index is rewritten to handle multiple segments.
-  private final LogSegment logSegment;
+  static final short version = 0;
+  static final String Index_File_Name_Suffix = "index";
```
Will change.

vgkholla replied:
Done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -1010,35 +1047,34 @@ public void write() throws StoreException {
-          long currentIndexEndOffsetBeforeFlush = currentInfo.getEndOffset();
-          long logEndOffsetBeforeFlush = logSegment.getEndOffset();
-          if (logEndOffsetBeforeFlush < currentIndexEndOffsetBeforeFlush) {
+          Offset currentIndexEndOffsetBeforeFlush = currentInfo.getEndOffset();
```
Will change.

vgkholla replied:
Actually, renaming this would make it inconsistent with how this is referred to in other places (`BlobStore.java`). I don't think the name is misleading either - any reason to suggest the change?

pnarayanan replied:
`current....BeforeFlush` sounds like an oxymoron.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexValue.java:
```diff
@@ -22,91 +22,95 @@
-    value.putLong(timeToLiveInMs);
+    value.putLong(expiresAtMs);
     value.putLong(originalMessageOffset);
     value.position(0);
```
No idea - Don't see the reason to do it either way and since this is not a part of my changes, will leave as is.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexValue.java:
```diff
@@ -22,91 +22,95 @@
-    value.putLong(Blob_Size_In_Bytes + Offset_Size_In_Bytes + Flag_Size_In_Bytes + Time_To_Live_Size_In_Bytes,
-        oldOffset);
+  void setNewOffset(Offset newOffset) {
+    long originalMessageOffset = offset.getName().equals(newOffset.getName()) ? offset.getOffset() : -1;
```
Not sure how that can be done?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/Log.java:
```diff
@@ -387,16 +384,26 @@ private void ensureCapacity(long writeSize) throws IOException {
-   * @return a {@link StoreMessageReadSet} with the file and file channel of the first segment and the given {@code }
-   * @deprecated this function is deprecated and is available for use until {@link PersistentIndex} and
-   * {@link HardDeleter} are rewritten to understand segmented logs.
+   * Gets the {@link FileSpan} for a message that is written starting at {@code endOffsetOfPrevMessage} and is of size
```
What is your suggestion for a more precise wording?

vgkholla replied:
Also note that end of prev segment == start of next segment logically.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -672,24 +643,63 @@ public boolean getEntriesSince(StoreKey key, FindEntriesCondition findEntriesCon
         logger.error("IndexSegment : " + indexFile.getAbsolutePath() + " index not found for key " + key);
       }
-    } else {
+    } else if (key == null || index.containsKey(key)) {
```
anything due to log segmentation in particular that warrants this change?

vgkholla replied:
No - but it was a bug that I discovered while writing tests and seemed easy enough to fix. Can revert if you think it is not required.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -21,87 +23,120 @@
   private static final short Version_Length = 2;
   private static final short Offset_Length = 8;
   private static final short Size_Length = 8;
   private static final short TTL_Length = 8;
```
`TTL_Length` should be renamed as well?

vgkholla replied:
Done.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java:
```diff
@@ -21,87 +23,120 @@
+        // backwards compatibility
+        Offset offset = new Offset(log.getFirstSegment().getName(), stream.readLong());
         long size = stream.readLong();
         long ttl = stream.readLong();
```
rename?

vgkholla replied:
Done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -126,13 +129,12 @@ public IndexSegment(String dataDir, long startOffset, StoreKeyFactory factory, i
    * @param journal The journal to use
    * @throws StoreException
    */
-  public IndexSegment(File indexFile, boolean isMapped, StoreKeyFactory factory, StoreConfig config,
```
Done

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java:
```diff
@@ -672,24 +643,63 @@ public boolean getEntriesSince(StoreKey key, FindEntriesCondition findEntriesCon
+   */
+  static Offset getIndexSegmentStartOffset(String filename) {
+    // file name pattern for index is logSegmentName_offset_index.
+    // If the logSegment name is empty, then the file name pattern is offset_index.
```
Done.

vgkholla commented:
@nsivabalan @pnarayanan Addressed comments. The commit that addresses the latest comments is 4246fee. The rest of the commit are due to rebasing to fix merge conflicts.

vgkholla commented:
@nsivabalan @pnarayanan Ping.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -798,7 +828,7 @@ private StoreFindToken findEntriesFromSegmentStartOffset(long initialSegmentStar
     }
 
-    if (newTokenOffsetInJournal != StoreFindToken.Uninitialized_Offset) {
+    if (newTokenOffsetInJournal != null) {
```
I don't get the purpose of this variable - when is this ever non-null?

vgkholla replied:
Discussed this offline - it is set just a few lines below but the line doesn't show up because it didn't need to be changed.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -1010,38 +1041,37 @@ public void write() throws StoreException {
+          Offset currentLogEndPointer = log.getEndOffset();
           while (prevInfo != null && !prevInfo.isMapped()) {
-            if (prevInfo.getEndOffset() > currentLogEndPointer) {
+            if (prevInfo.getEndOffset().compareTo(currentLogEndPointer) > 0) {
```
Isn't this the bug you discovered? Why do we persist the latest segment and then backwards?

vgkholla replied:
Yup, this is the bug. I haven't changed it because it is a big enough change on its own. Have created #501 for this

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -1010,38 +1041,37 @@ public void write() throws StoreException {
         if (indexes.size() > 0) {
           // before iterating the map, get the current file end pointer
-          Map.Entry<Long, IndexSegment> lastEntry = indexes.lastEntry();
+          Map.Entry<Offset, IndexSegment> lastEntry = indexes.lastEntry();
```
With compaction, did you say you are going to embed `indexes` (or rather `segments` as this should be named) inside an `AtomicReference`? Because otherwise, this persistence is going to be tricky and you would have to make use of an iterator to traverse.

In fact, this code should really have been using an iterator over the segments rather than going with this `lowerEntry` business. I understand that you don't want to make extensive changes now so it is fine - but you might want to revisit this during compaction.

vgkholla replied:
Yes, there is going to be an `AtomicReference` to make sure the persistence works correctly.

pnarayanan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -1071,8 +1102,15 @@ public void run() {
+   * The type of the store token.
+   */
+  enum Type {
+    Uninitialized, JournalBased, IndexBased
```
+1. This was a much needed change regardless of segmentation.

vgkholla commented:
@pnarayanan Review comments addressed.

pnarayanan commented:
LGTM. @nsivabalan please take a look at the recent changes and feel free to merge if they look alright.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java:
```diff
@@ -53,14 +53,14 @@
    */
   private enum IndexEntryType {
-    PUT, DELETE,
+    ANY, PUT, DELETE,
```
minor: can we have documentation for these types
